File: mcp-client/server_connection.py
# mcp-client/server_connection.py

import time
from typing import Dict, List, Optional, Any
from contextlib import AsyncExitStack
from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client
from mcp.client.sse import sse_client

class ServerConnection:
    def __init__(self):
        self.sessions: Dict[str, ClientSession] = {}
        self.exit_stack = AsyncExitStack()
        self.server_info: Dict[str, Dict] = {}

    async def connect_to_servers(self, servers: Dict[str, Dict[str, Any]]) -> None:
        """根据配置文件连接到所有 MCP 服务器"""
        if not servers:
            print("No servers found in configuration.")
            return

        for server_name, server_config in servers.items():
            command = server_config.get("command")
            args = server_config.get("args", [])

            if not command or not isinstance(args, list):
                print(f"Invalid configuration for server {server_name}: {server_config}")
                continue

            if command.lower() == "sse":
                transport_type = "sse"
                if not args or not args[0].startswith("http"):
                    print(f"Invalid SSE URL for server {server_name}: {args}")
                    continue
                url = args[0]
                address = url.split("://")[1].split(":")[0]
                port = int(url.split(":")[-1].split("/")[0])
            else:
                transport_type = "stdio"
                address = args[0] if args else ""
                port = 0

            print(f"Connecting to server: {server_name} at {address}:{port} (Transport: {transport_type})")
            await self.connect_to_server(server_name, address, port, transport_type, command, args)

    async def connect_to_server(self, server_name: str, address: str, port: int, transport_type: str, command: str, args: List[str]) -> None:
        """连接到 MCP 服务器（支持 stdio 和 sse 模式）"""
        try:
            if transport_type == "stdio":
                server_params = StdioServerParameters(
                    command=command,
                    args=args,
                    env=None
                )
                stdio_transport = await self.exit_stack.enter_async_context(stdio_client(server_params))
                stdio, write = stdio_transport
                session = await self.exit_stack.enter_async_context(ClientSession(stdio, write))

            elif transport_type == "sse":
                url = args[0]
                sse_transport = await self.exit_stack.enter_async_context(sse_client(url))
                if isinstance(sse_transport, tuple):
                    read, write = sse_transport
                    class SSETransportAdapter:
                        def __init__(self, read, write):
                            self.read = read
                            self.write = write
                    transport = SSETransportAdapter(read, write)
                else:
                    transport = sse_transport
                session = await self.exit_stack.enter_async_context(ClientSession(transport.read, transport.write))

            else:
                raise ValueError(f"Unsupported transport type: {transport_type}")

            await session.initialize()
            start_time = time.time()
            tools_response = await session.list_tools()
            resources_response = await session.list_resources()
            prompts_response = await session.list_prompts()
            latency = (time.time() - start_time) * 1000

            # 存储工具的详细信息，包括 inputSchema
            tools = [
                {
                    "name": tool.name,
                    "description": tool.description,
                    "inputSchema": getattr(tool, "inputSchema", {}),
                }
                for tool in tools_response.tools
            ]

            # 存储资源的详细信息
            resources = [
                {
                    "uri": resource.uri,
                    "mimeType": resource.mimeType,
                    "name": getattr(resource, "name", ""),
                    "description": getattr(resource, "description", ""),
                }
                for resource in resources_response.resources
            ]

            # 存储 prompt 的详细信息，将 PromptArgument 转换为字典
            prompts = []
            for prompt in prompts_response.prompts:
                arguments = [
                    {
                        "name": arg.name,
                        "description": arg.description,
                        "required": arg.required,
                    }
                    for arg in getattr(prompt, "arguments", [])
                ]
                prompts.append({
                    "name": prompt.name,
                    "description": prompt.description,
                    "arguments": arguments,
                })

            # 获取每个 prompt 的模板
            prompt_templates = {}
            for prompt in prompts:
                prompt_name = prompt["name"]
                prompt_response = await session.get_prompt(prompt_name)
                prompt_templates[prompt_name] = {
                    "description": prompt_response.description,
                    "messages": [
                        {
                            "role": msg.role,
                            "content": msg.content.text if msg.content.type == "text" else "",
                        }
                        for msg in prompt_response.messages
                    ],
                }

            self.sessions[server_name] = session
            self.server_info[server_name] = {
                "tools": tools,
                "resources": resources,
                "prompts": prompts,
                "prompt_templates": prompt_templates,
                "latency": latency,
                "load": 0.0,
            }

            print(f"\nConnected to {server_name} with tools: {[tool['name'] for tool in tools]}, latency: {latency:.2f}ms")

        except Exception as e:
            print(f"Failed to connect to {server_name}: {str(e)}")
            if server_name in self.sessions:
                del self.sessions[server_name]
            if server_name in self.server_info:
                del self.server_info[server_name]

    async def cleanup(self):
        """清理资源"""
        await self.exit_stack.aclose()

File: mcp-client/src/infra/serverConnection.ts
// mcp-client/src/infra/serverConnection.ts

import { Client, ClientOptions } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import { SSEClientTransport } from "@modelcontextprotocol/sdk/client/sse.js";


interface ClientInfo {
  name: string;
  version: string;
  [key: string]: unknown; // Add index signature
}

interface ServerConfig {
  command: string;
  args: string[];
}

interface Tool {
  name: string;
  description: string;
  inputSchema: any;
}

interface Resource {
  uri: string;
  mimeType: string;
  name: string;
  description: string;
}

interface PromptArgument {
  name: string;
  description: string;
  required: boolean;
}

interface Prompt {
  name: string;
  description: string;
  arguments: PromptArgument[];
}

interface PromptMessage {
  role: string;
  content: { type: string; text: string };
}

interface ServerInfo {
  tools: Tool[];
  resources: Resource[];
  prompts: Prompt[];
  promptTemplates: Record<string, { description: string; messages: PromptMessage[] }>;
  latency: number;
  load: number;
}

export class ServerConnection {
  private clients: Map<string, Client> = new Map();
  public serverInfo: Map<string, ServerInfo> = new Map();

  async connectToServers(servers: Record<string, ServerConfig>): Promise<void> {
    if (!Object.keys(servers).length) {
      console.log("No servers found in configuration.");
      return;
    }

    for (const [serverName, serverConfig] of Object.entries(servers)) {
      const { command, args } = serverConfig;

      if (!command || !Array.isArray(args)) {
        console.log(`Invalid configuration for server ${serverName}: ${JSON.stringify(serverConfig)}`);
        continue;
      }

      let transportType: "stdio" | "sse";
      let address: string;
      let port: number;

      if (command.toLowerCase() === "sse") {
        transportType = "sse";
        if (!args.length || !args[0].startsWith("http")) {
          console.log(`Invalid SSE URL for server ${serverName}: ${args}`);
          continue;
        }
        const url = args[0];
        address = url.split("://")[1].split(":")[0];
        const portPart = url.split(":").slice(-1)[0]?.split("/")[0];
        port = portPart ? parseInt(portPart) : 0;
      } else {
        transportType = "stdio";
        address = args[0] || "";
        port = 0;
      }

      console.log(`Connecting to server: ${serverName} at ${address}:${port} (Transport: ${transportType})`);
      await this.connectToServer(serverName, transportType, command, args);
    }
  }

  async connectToServer(serverName: string, transportType: "stdio" | "sse", command: string, args: string[]): Promise<void> {
    try {
      const clientInfo: ClientInfo = {
        name: `mcp-client-${serverName}`,
        version: "1.0.0",
      };
      const clientOptions: ClientOptions = {
        capabilities: {
          prompts: {},
          resources: {},
          tools: {},
        },
      };
      const client = new Client(clientInfo, clientOptions);

      if (transportType === "stdio") {
        const transport = new StdioClientTransport({ command, args });
        await client.connect(transport);
      } else {
        const url = args[0];
        const transport = new SSEClientTransport(new URL(url));
        await client.connect(transport);
      }

      const startTime = Date.now();

      const toolsResponse = await client.listTools();
      const resourcesResponse = await client.listResources();
      const promptsResponse = await client.listPrompts();
      const latency = Date.now() - startTime;

      const tools = toolsResponse.tools.map((tool: { name: string; description?: string; inputSchema: any }) => ({
        name: tool.name,
        description: tool.description || "",
        inputSchema: tool.inputSchema || {},
      }));

      const resources = resourcesResponse.resources.map(
        (resource: { uri: string; mimeType?: string; name: string; description?: string }) => ({
          uri: resource.uri,
          mimeType: resource.mimeType || "application/octet-stream",
          name: resource.name,
          description: resource.description || "",
        })
      );

      const prompts = promptsResponse.prompts.map(
        (prompt: { name: string; description?: string; arguments?: Array<{ name: string; description?: string; required?: boolean }> }) => ({
          name: prompt.name,
          description: prompt.description || "",
          arguments: (prompt.arguments || []).map((arg: { name: string; description?: string; required?: boolean }) => ({
            name: arg.name,
            description: arg.description || "",
            required: arg.required || false,
          })),
        })
      );

      const promptTemplates: Record<string, { description: string; messages: PromptMessage[] }> = {};
      for (const prompt of prompts) {
        const promptName = prompt.name;
        const promptResponse = await client.getPrompt({ name: promptName });
        promptTemplates[promptName] = {
          description: promptResponse.description||"",
          messages: (promptResponse.messages || []).map((msg: { role: "user" | "assistant"; content: { type: string; text?: string; data?: string; mimeType?: string } }) => ({
            role: msg.role,
            content: msg.content.type === "text" && msg.content.text
              ? { type: "text", text: msg.content.text }
              : { type: "text", text: `[Unsupported content type: ${msg.content.type}]` },
          })),
        };
      }

      this.clients.set(serverName, client);
      this.serverInfo.set(serverName, {
        tools,
        resources,
        prompts,
        promptTemplates,
        latency,
        load: 0.0,
      });

      console.log(`\nConnected to ${serverName} with tools: ${tools.map((tool: { name: string }) => tool.name)}, latency: ${latency}ms`);
    } catch (error) {
      console.log(`Failed to connect to ${serverName}: ${error}`);
      this.clients.delete(serverName);
      this.serverInfo.delete(serverName);
    }
  }

  getSession(serverName: string): Client | undefined {
    return this.clients.get(serverName);
  }

  async cleanup(): Promise<void> {
    for (const [serverName, client] of this.clients.entries()) {
      await client.close();
      this.clients.delete(serverName);
    }
  }
}